lib/dfu_target_uart_host/dfu_target_uart_host.c

```diff
@@ -0,0 +1,326 @@
+
+	return 0;
+}
+
+static int parse_done(const struct DFU_UART_target_RemoteFnc *fnc)
+{
+	context.download_success = fnc->data.success;
+
+	return 0;
+}
+
+static void execute_input_fnc(const struct DFU_UART_target_RemoteFnc *fnc)
+{
+	int err;
+	size_t offs_get;
+	static size_t progress;
+
+	LOG_DBG("DFU function called: %d", fnc->type_choice);
+
+	switch (fnc->type_choice) {
+	case _DFU_UART_target_RemoteFnc_type_init:
+		parse_init(fnc);
+		progress = 0;
+		err = dfu_target_mcuboot_init(context.file_size,
+					      context.image_num,
+					      NULL);
+		send_error(
+			_DFU_UART_target_RemoteFnc_type_init,
+			err);
+		break;
+	case _DFU_UART_target_RemoteFnc_type_write:
+		parse_write(fnc);
+		err = dfu_target_mcuboot_write(context.to_write,
+					       context.to_write_len);
+		progress += context.to_write_len;
+		LOG_INF("Progress: %d %%", progress * 100 / context.file_size);
+		send_error(
+			_DFU_UART_target_RemoteFnc_type_write,
+			err);
+		break;
+	case _DFU_UART_target_RemoteFnc_type_offset:
+		err = dfu_target_mcuboot_offset_get(&offs_get);
+		send_offset(err, offs_get);
+		break;
+	case _DFU_UART_target_RemoteFnc_type_done:
+		parse_done(fnc);
+		err = dfu_target_mcuboot_done(context.download_success);
+		if (err) {
+			send_error(
+				_DFU_UART_target_RemoteFnc_type_done, err);
+			break;
+		}
+
+		err = dfu_target_mcuboot_schedule_update(
+			context.image_num);
+		send_error(
+			_DFU_UART_target_RemoteFnc_type_done,
+			err);
+		break;
+	default:
+		break;
+	}
+}
+
+static void process_input_data(void)
+{
+	int err;
+	uint8_t *cursor;
+	size_t cursor_len;
+	size_t dec_len = 0;
+	struct DFU_UART_target_RemoteFnc fnc = {0};
+
+	while (true) {
+		cursor_len = ring_buf_get_claim(&context.rx_rb, &cursor, UINT32_MAX);
+		if (cursor_len == 0) {
+			/* Let's reset it to prevent wrapping */
+			ring_buf_reset(&context.rx_rb);
+			return;
+		}
+
+		/* We try to put the whole buffer in the decode function.
+		 * If there is not enough data, the decode will throw an error.
+		 * If there is too much data, no error will be thrown and
+		 * the correct amount of decoded data will be given by the
+		 * decode function
+		 */
+		err = cbor_decode_DFU_UART_target_RemoteFnc(
+			cursor,
+			cursor_len,
+			&fnc,
+			&dec_len);
+		if (err) {
+			ring_buf_get_finish(&context.rx_rb, 0);
+			return;
+		}
+
+		execute_input_fnc(&fnc);
+
+		/* We need to remove the amount of consumed data */
+		err = ring_buf_get_finish(&context.rx_rb, dec_len);
+		if (err) {
+			LOG_ERR("ring_buf_get_finish returned %d", err);
+		}
+	}
+}
+
+noreturn static void uart_input_handler(void *p1, void *p2, void *p3)
+{
+	ARG_UNUSED(p1);
+	ARG_UNUSED(p2);
+	ARG_UNUSED(p3);
+
+	while (1) {
+		k_sem_take(&context.lte_uart_sem, K_FOREVER);
+		process_input_data();
+	}
+}
+
+int dfu_target_uart_host_start(const struct device *uart)
+{
+	if (context.init_done) {
+		return -EAGAIN;
+	}
+
+	if (!device_is_ready(uart)) {
+		LOG_ERR("UART in not valid");
```

Comment:
This is only true when `uart` is NULL.
Is logging really needed here, can just errcode be returned and logging left for caller?